WcaOnRails/app/controllers/accounts/registrations_controller.rb

```diff
@@ -1,7 +1,20 @@
 # frozen_string_literal: true
 
 class Accounts::RegistrationsController < Devise::RegistrationsController
+  prepend_before_action :check_captcha, only: [:create]
```

Comment:
```suggestion
  before_action :check_captcha, only: [:create]
```

---

We are permitting a specific list of params in a `before_action` in the base controller:

https://github.com/thewca/worldcubeassociation.org/blob/215540727b15242896029663fa3455b04f06cf7e/WcaOnRails/app/controllers/application_controller.rb#L54-L65

If we use `prepend_before_action` here, it means we run `check_captcha` first, and only then `configure_permitted_parameters`, which is definitely not what we want. Consequently, on captcha failure all those params would be ignored. @gregorbg that's the reason why the form gets cleared.

Reply:
This makes a lot of sense, especially with the code snippet that you referenced. But to be honest it also feels a little arcane that there's such a significant difference between `prepend_before_action` and `before_action` which manifests itself in the form being cleared. Anyways, thanks for spotting!

Reply:
Thanks, @jonatanklosko for spotting it. All code snippets regarding captcha had default implementation of preprend_before_action on the internet, so didn't cross my mind that this would be the culprit.